filter_plugins/list-dict-diff.py

```diff
@@ -8,9 +8,62 @@ def get_remaining_items(list_a, list_b, key_a, key_b):
 
     return list_a
 
+# For manage-workflow-templates role
+# Takes a workflow defined in a dictionary and processes it into a list
+def flatten_workflow_nodes(workflows):
+    flattened_workflow, flattened_start_tree = [], []
+    current_node_index = 1
+
+    # Processes each node beginning at start and all their success/failure branches
+    for workflow in workflows:
+        flattened_start_tree = gen_flattened_list(workflow, current_node_index, 0, 'start', [])
+        current_node_index += len(flattened_start_tree)
+        flattened_workflow = flattened_workflow + flattened_start_tree
+
+    return flattened_workflow
+
+# Helper function for flatten_workflow_nodes
+# Recursively flattens a workflow dictionary with one node at start
+def gen_flattened_list(workflow, node_index, parent_index, node_type, flattened_nodes):
+    current_node = [{
+     'job_name': workflow['unified_job_template']['name'],
+     'type': node_type,
+     'index': node_index,
+     'parent_index': parent_index
+   }]
+
+   # Return node it has no children
+    if('success_nodes' not in workflow and 'failure_nodes' not in workflow or
+        (not workflow['success_nodes'] and not workflow['failure_nodes'])):
+            return current_node
+
+    # Add current node to the list of processed nodes
+    flattened_nodes = flattened_nodes + current_node
+    child_index = node_index + 1
+
+    # Recursively process all of the current nodes success children
```

Comment:
The `success_nodes` and `failure_nodes` have the exact same processing (with the exception of the `node_type`). Let's attempt to combine to avoid dup code - i.e.: just a quick little loop for two elements would do - i.e.: for `success_nodes` and `failure_nodes`.

Reply:
i.e.: something like:
```
for node_type in [ "success_nodes", "failure_nodes" ]:
    if (node_type in workflow and len(workflow[node_type]) > 0):
        new_nodes = []
        for node in workflow['success_nodes']:
            new_nodes = gen_flattened_list(node, child_index, node_index, node_type, [])
            child_index += len(new_nodes)
            flattened_nodes = flattened_nodes + new_nodes
```

Note how the parameter to `gen_flattened_list` also uses `node_type` - more on that below.